Sources/Core/Sources/ECS/Systems/PlayerInputSystem.swift

```diff
@@ -136,16 +136,49 @@ public final class PlayerInputSystem: System {
       if event.key == .enter {
         if !message.isEmpty {
           try connection?.sendPacket(ChatMessageServerboundPacket(message))
+          guiState.playerMessageHistory.append(message)
+          guiState.currentMessageIndex = nil
         }
         guiState.messageInput = nil
         return true
       } else if event.key == .escape {
         guiState.messageInput = nil
+        guiState.currentMessageIndex = nil
         return true
       } else if event.key == .delete {
         if !message.isEmpty {
           guiState.messageInput?.removeLast()
         }
+      } else if event.key == .upArrow {
+        // If no message is selected, select the above message
+        if let index = guiState.currentMessageIndex, index > 0 {
+          guiState.currentMessageIndex = index - 1
+          guiState.messageInput = guiState.playerMessageHistory[index - 1]
+        } else if guiState.currentMessageIndex == nil {
+          guiState.stashedMessageInput = guiState.messageInput
+          if !guiState.playerMessageHistory.isEmpty {
+            guiState.currentMessageIndex = guiState.playerMessageHistory.count - 1
+            if let index = guiState.currentMessageIndex {
+              guiState.messageInput = guiState.playerMessageHistory[index]
+            }
```

Comment:
You can actually avoid the `if let` by replacing lines 160-163 with the following snippet. This is better because in your code the `if` actually makes it look like there's conditional logic when there's not (it will always execute).

```swift
let index = guistate.playerMessageHistory.count - 1
guiState.currentMessageIndex = index
guiState.messageInput = guiState.playerMessageHistory[index]
```